doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,189 @@
+# Diagnostics Channel
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+<!-- source_link=lib/diagnostics_channel.js -->
+
+The `diagnostics_channel` module provides an API to report diagnostic data
+via named channels. It can be accessed using:
```

Comment:
Anything. It's intentionally an unspecified bag of data. It's up to each publisher to decide what should be in the bag of data for their given named channel. It could be span data expressing timing information for some block of execution, it could be a interval reported metrics, it could be details about a request like hostname, etc. It's meant to be a generic channel that APM vendors can have core and userland modules emit the data they need from the various places they need to collect data from, and what that data is varies significantly between different parts of the application.

On top of this generic channel, I intend to create some generic common constructs to report certain things consistently, like spans and metrics.

Reply:
With regards to metrics... How, if at all, does this relate to the `perf_hooks` module?

Reply:
same for https://nodejs.org/api/tracing.html

Reply:
It can be used to send data similar to what perf_hooks produces, but is just a delivery mechanism, it could send _anything_. We _could_ probably make some integration to get `perf_hooks` to emit events to its own channel in addition to the observer or something like that, if we wanted.

As for `trace_events`, it overlaps in that it is a data firehose, but is different in that it can be consumed within JavaScript and allows sending of arbitrary data which makes it a lot more useful for APMs. While `trace_event` is great for post-mortem diagnostics, it's not really suitable for production monitoring and in particular is not delivered in a format that it can be tied adequately to the execution context to produce a request trace, only a process trace.